File: decorator/go/decorator.go
package decorator

import (
	"bytes"
	"compress/zlib"
	"encoding/base64"
	"fmt"
	"io"
)

type DataSource interface {
	Read() string
	Write(data string)
}

type FileDataSource struct {
	Name        string
	fileContent string
}

func NewFileDataSource(name string) *FileDataSource {
	return &FileDataSource{Name: name}
}

func (f *FileDataSource) Read() string {
	fmt.Printf("FileDataSource 读取文件 %s\n", f.Name)
	return f.fileContent
}

func (f *FileDataSource) Write(data string) {
	f.fileContent = data
	fmt.Printf("FileDataSource 写入到文件 %s 中，数据为：%s\n", f.Name, data)
}

// ============================================================

type DataSourceDecorator struct {
	wrapper DataSource
}

func (d *DataSourceDecorator) Read() string {
	return d.wrapper.Read()
}

func (d *DataSourceDecorator) Write(data string) {
	d.wrapper.Write(data)
}

type EncryptionDecorator struct {
	*DataSourceDecorator
}

func NewEncryptionDecorator(source DataSource) *EncryptionDecorator {
	return &EncryptionDecorator{&DataSourceDecorator{source}}
}

func (e *EncryptionDecorator) Read() string {
	decoded, _ := base64.StdEncoding.DecodeString(e.wrapper.Read())
	for i := 0; i < len(decoded); i++ {
		decoded[i] -= 1
	}
	return string(decoded)
}

func (e *EncryptionDecorator) Write(data string) {
	result := []byte(data)
	for i := 0; i < len(result); i++ {
		result[i] += 1
	}
	e.wrapper.Write(base64.StdEncoding.EncodeToString(result))
}

type CompressionDecorator struct {
	CompLevel int
	*DataSourceDecorator
}

func NewCompressionDecorator(source DataSource) *CompressionDecorator {
	return &CompressionDecorator{
		CompLevel:           zlib.DefaultCompression,
		DataSourceDecorator: &DataSourceDecorator{source},
	}
}

func (c *CompressionDecorator) Read() string {
	decompressed, _ := c.decompress(c.wrapper.Read())
	return decompressed
}

func (c *CompressionDecorator) Write(data string) {
	compressed, _ := c.compress(data)
	c.wrapper.Write(compressed)
}

func (c *CompressionDecorator) compress(src string) (target string, err error) {
	input := []byte(src)

	var compressed bytes.Buffer
	writer, err := zlib.NewWriterLevel(&compressed, c.CompLevel)
	if err != nil {
		return
	}

	_, err = writer.Write(input)
	if err != nil {
		_ = writer.Close()
		return
	}

	// 注意这里不能用 defer，因为要保证 base64.StdEncoding.EncodeToString(...) 在 writer.Close() 之后执行
	err = writer.Close()
	target = base64.StdEncoding.EncodeToString(compressed.Bytes())
	return
}

func (c *CompressionDecorator) decompress(src string) (target string, err error) {
	decoded, err := base64.StdEncoding.DecodeString(src)
	if err != nil {
		return
	}

	reader, err := zlib.NewReader(bytes.NewReader(decoded))
	if err != nil {
		return
	}

	defer func() {
		e := reader.Close()
		if err == nil {
			err = e
		}
	}()

	var decompressed bytes.Buffer
	_, err = io.Copy(&decompressed, reader)
	if err != nil {
		return
	}
	target = decompressed.String()
	return
}

File: decorator/java/Decorator.java
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Base64;
import java.util.zip.Deflater;
import java.util.zip.DeflaterOutputStream;
import java.util.zip.InflaterInputStream;

/**
 * 定义了读取和写入操作的通用数据接口
 */
interface DataSource {
    String read();

    void write(String data);
}

/**
 * 简单数据读写器
 */
class FileDataSource implements DataSource {
    private final String name;
    /**
     * 写入到文件的内容（这里是模拟，并没有真正写入到文件中）
     */
    private String fileContent = "";

    public FileDataSource(String name) {
        this.name = name;
    }

    @Override
    public String read() {
        System.out.println("FileDataSource 读取文件 " + name);
        return fileContent;
    }

    @Override
    public void write(String data) {
        fileContent = data; // 模拟写入文件中
        System.out.println("FileDataSource 写入到文件 " + name + " 中，数据为：" + data);
    }
}

// ============================================================

/**
 * 抽象基础装饰，这里实现 DataSource 是为了可以给其他装饰器再次装饰（甚至可以递归装饰）
 */
class DataSourceDecorator implements DataSource {
    private final DataSource delegate;

    public DataSourceDecorator(DataSource source) {
        this.delegate = source;
    }

    @Override
    public String read() {
        return delegate.read();
    }

    @Override
    public void write(String data) {
        delegate.write(data);
    }
}

/**
 * 具体装饰实现 - 加解密装饰
 */
class EncryptionDecorator extends DataSourceDecorator {
    public EncryptionDecorator(DataSource source) {
        super(source);
    }

    @Override
    public String read() {
        return decode(super.read());
    }

    @Override
    public void write(String data) {
        super.write(encode(data));
    }

    private String encode(String data) {
        byte[] result = data.getBytes();
        for (int i = 0; i < result.length; i++) {
            result[i] += (byte) 1;
        }
        return Base64.getEncoder().encodeToString(result);
    }

    private String decode(String data) {
        byte[] result = Base64.getDecoder().decode(data);
        for (int i = 0; i < result.length; i++) {
            result[i] -= (byte) 1;
        }
        return new String(result);
    }
}

/**
 * 具体装饰实现 - 压缩/解压装饰
 */
class CompressionDecorator extends DataSourceDecorator {
    private int compLevel = Deflater.DEFAULT_COMPRESSION;

    public CompressionDecorator(DataSource source) {
        super(source);
    }

    public int getCompressionLevel() {
        return compLevel;
    }

    public void setCompressionLevel(int value) {
        compLevel = value;
    }

    @Override
    public void write(String data) {
        super.write(compress(data));
    }

    @Override
    public String read() {
        return decompress(super.read());
    }

    private String compress(String src) {
        byte[] bytesData = src.getBytes();
        try {
            // 这里的 OutputStream、ByteArrayOutputStream、DeflaterOutputStream 其实也是用了装饰器模式
            // 注意这里不能使用 try-with-resources 语法，否则会报 java.io.EOFException: Unexpected end of ZLIB input stream
            ByteArrayOutputStream bout = new ByteArrayOutputStream(512);
            DeflaterOutputStream dos = new DeflaterOutputStream(bout, new Deflater(compLevel));
            dos.write(bytesData);
            dos.close();
            bout.close();
            return Base64.getEncoder().encodeToString(bout.toByteArray());
        } catch (IOException ignore) {
            return null;
        }
    }

    private String decompress(String src) {
        byte[] decodedData = Base64.getDecoder().decode(src);
        try {
            // 这里的 InputStream、ByteArrayInputStream、InflaterInputStream 其实也是用了装饰器模式
            // 注意这里不能使用 try-with-resources 语法，否则会报 java.io.EOFException: Unexpected end of ZLIB input stream
            InputStream in = new ByteArrayInputStream(decodedData);
            InflaterInputStream iin = new InflaterInputStream(in);
            ByteArrayOutputStream bout = new ByteArrayOutputStream(512);
            int b;
            while ((b = iin.read()) != -1) {
                bout.write(b);
            }
            bout.close();
            iin.close();
            in.close();
            return bout.toString();
        } catch (IOException ignore) {
            return null;
        }
    }
}

// 可以有更多的具体装饰实现...